### PR TITLE
TST: (numpy dev) fix return type for a selection of ufuncs applied to Column

### LINKED ISSUE #15926: TST: (numpy dev) new failing tests
### Description

Three tests appear to be failing in devdeps even after #15924 is dealt with. Namely
- `astropy/units/tests/test_quantity_non_ufuncs.py::TestArgFunctions::test_argsort`
- `astropy/utils/masked/tests/test_function_helpers.py::TestArgFunctions::test_argsort`
- `astropy/utils/masked/tests/test_function_helpers.py::TestSortFunctions::test_sort`

see https://github.com/astropy/astropy/actions/runs/7609867763/job/20722001589

~I note that I'm unable to reproduce these failures locally (all tests pass with numpy built from source), so either the problem is already addressed upstream and wheels are just a couple hours behind (in which case it'll resolve itself tomorrow), or this is somehow platform specific and doesn't affect macOS.~
edit: I don't know what I did wrong that day but in any case that's now incorrect, and I can now reproduce locally.

### Expected behavior

_No response_

### How to Reproduce

Install numpy dev, build pyerfa + astropy against it, then run (for instance)
```shell
pytest astropy/units/tests/test_quantity_non_ufuncs.py::TestArgFunctions::test_argsort
```

### Versions

```python
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import astropy; print("astropy", astropy.__version__)
import numpy; print("Numpy", numpy.__version__)
import erfa; print("pyerfa", erfa.__version__)
#import scipy; print("Scipy", scipy.__version__)
#import matplotlib; print("Matplotlib", matplotlib.__version__)
```
```
macOS-14.2.1-arm64-arm-64bit
Python 3.12.1 (main, Dec 11 2023, 18:41:50) [Clang 15.0.0 (clang-1500.0.40.1)]
astropy 6.1.dev350+g681436af1d
Numpy 2.0.0.dev0+git20240122.89eac8b
pyerfa 2.0.1.2.dev11+gc498084
```

edit: PRs opened to fix this
- #15929
- #15930 
- #15941

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

This pull request is to address part of #15926, specifically fixing `astropy/table/tests/test_column.py::TestColumn::test_numpy_boolean_ufuncs` (as discussed in https://github.com/astropy/astropy/issues/15926#issuecomment-1907681895)

~The fix is rather inelegant, but I'm not sure there's a better one.~

Close #15926

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I think this is working, but for extra safety, I'll wait until #15929  is merged so I can rebase this one and hopefully see the number of failure going *down* (I expect 16 remaining failures after this patch, which is 2 less than what is currently seen in #15929)

**Comment by neutrinoceros** [comment]
Actually rebased onto #15929 so it's clearer what this PR brings.

**Comment by neutrinoceros** [comment]
following the success of my experimental merger (#15949), let's rebase on main again so merging order can be left to reviewers.

File: astropy/table/column.py
Comment (mhvk) [inline astropy/table/column.py:754]: this section can be
```
if isinstance(out_arr, BaseColumn):
    out_arr = out_arr.data
```
Actually, never mind... see next comment

File: astropy/table/column.py
Comment (mhvk) [inline astropy/table/column.py:760]: Actually, the problem is that the previous fix caused a breakage: `np.isfinite` is in `_comparison_functions`. So, we need to combine the two, as follows:
```
if self.shape != out_arr.shape or (
        isinstance(out_arr, BaseColumn)
        and (context is not None and context[0] in _comparison_functions)
    ):
    data = out_arr.data
    return data[()] if NUMPY_LT_2_0 or return_scalar else data
else:
    return out_arr
```

**Review by mhvk — COMMENTED** [review]
Looks like the previous fix we had actually made other things worse... Here is a new suggestion...

**Comment by neutrinoceros** [comment]
> Looks like the previous fix we had actually made other things worse...

What do you mean ? Is it because the number of failure is greater on this branch than you expect ? I'm going to rebase now, which *should* bring the number of failures to zero. While CI runs I'll also read your other comments.

File: astropy/table/column.py
Comment (neutrinoceros) [reply astropy/table/column.py:760]: I'm not sure what breakage you've identified, and see no failing test with this branch (after rebase) locally.
For completion, I *also* don't see any failures with your suggestion applied.
I'll let CI complete here to see if I'm missing anything.

Comment (neutrinoceros) [reply astropy/table/column.py:760]: Ok, all I see is ✅, and it's night time in France, so I'm logging off, but feel free to push your suggestion to my branch if needed ! Thank you !

**Comment by mhvk** [comment]
Well, not worse, but broke something that worked before. Anyway, the fix should now be easy -- I pushed a revised version up.

**Review by mhvk — APPROVED** [review]
Should be good to go if tests pass!

**Comment by neutrinoceros** [comment]
@mhvk thank you. I'll be in reply-only until Wednesday, so please feel free to patch up my PRs in the mean time. Also note that auto-merge won't work while CI is unstable (see #15960).

**Comment by pllim** [comment]
Auto merge didn't work due to pytest 8 (sigh, the timing!). If the failures here are consistent with that I see in main with pytest 8, then I will merge.

**Comment by pllim** [comment]
Thanks, all!

**Pull request merged**
